ballerina-shell/modules/shell-cli/src/main/java/io/ballerina/shell/cli/jline/parser/ParserStateMachine.java

```diff
@@ -0,0 +1,317 @@
+    private static final char OPEN_CURLY = '{';
+    private static final char CLOSE_CURLY = '}';
+    private static final char OPEN_PAREN = '(';
+    private static final char CLOSE_PAREN = ')';
+    private static final char OPEN_SQ_BR = '[';
+    private static final char CLOSE_SQ_BR = ']';
+    private static final char BACKWARD_SLASH = '\\';
+    private static final char FORWARD_SLASH = '/';
+    private static final char HASH = '#';
+    private static final char DOLLAR = '$';
+    private static final char DOUBLE_QUOTES = '\"';
+    private static final char NEW_LINE = '\n';
+    private static final char CARRIAGE_RETURN = '\r';
+    private static final char SPACE = ' ';
+    private static final char TAB = '\t';
+    // Continuing operator characters are characters than cannot
+    // end a syntactically correct snippet: specifically :@*%+-<=&|^!,
+    // Note that you cannot include,
+    // . because floating point literals can end with a .
+    // > because x is map<int> is a valid expression (also XML filter expressions can end with >)
+    // '?' because x is int? is a valid expression
+    // '/' will be handled by AFTER_FORWARD_SLASH so it cannot be included here
+    private static final Set<Character> CONTINUING_OPERATORS =
+            Set.of(':', '@', '*', '%', '+', '-', '<', '=', '&', '|', '^', '!', ',');
+    private static final Map<Character, Character> OPEN_BRACKETS =
+            Map.of(CLOSE_CURLY, OPEN_CURLY,
+                    CLOSE_PAREN, OPEN_PAREN,
+                    CLOSE_SQ_BR, OPEN_SQ_BR);
+
+    private final Stack<Character> stack;
+    private ParserState state;
+
+    public ParserStateMachine() {
+        this.state = ParserState.NORMAL;
+        this.stack = new Stack<>();
+    }
+
+    public void feed(char character) {
+        if (character == CARRIAGE_RETURN) {
+            // Ignore carriage return
+            return;
+        }
+
+        switch (state) {
+            case NORMAL:
+                normalState(character);
+                break;
+            case AFTER_BACKWARD_SLASH:
+                afterBackwardSlashState();
+                break;
+            case IN_DOUBLE_QUOTES:
+                inDoubleQuotesState(character);
+                break;
+            case IN_DOUBLE_QUOTES_AFTER_BACKWARD_SLASH:
+                inDoubleQuotesAfterBackwardSlashState(character);
+                break;
+            case IN_TEMPLATE:
+                inTemplateState(character);
+                break;
+            case IN_TEMPLATE_AFTER_DOLLAR:
+                inTemplateAfterDollarState(character);
+                break;
+            case AFTER_FORWARD_SLASH:
+                afterForwardSlashState(character);
+                break;
+            case IN_COMMENT:
+                inCommentState(character);
+                break;
+            case AFTER_OPERATOR:
+                afterOperatorState(character);
+                break;
+            default:
```

Comment:
We don't need this, right? Or do need to do something here for other states (if any).

Reply:
The default case? spotbugs complained when I didn't include a default case, so thats why I added it.